File: conan/packages/glaze/glaze/doc_parser.py
"""Parser for OpenGL-Refpages DocBook XML documentation.

Extracts brief descriptions and parameter documentation from the Khronos
OpenGL reference pages and provides them as a simple lookup index.
"""

import os
import xml.dom.minidom
from dataclasses import dataclass, field
from xml.dom.minidom import Node


@dataclass
class FunctionDoc:
    """Documentation for a single GL function."""

    brief: str
    params: dict[str, str] = field(default_factory=dict)


# Mapping from Glaze API name to the refpages subdirectory that best covers it.
_API_TO_REFDIR: dict[str, str] = {
    "gl": "gl4",
    "gles1": "es1.1",
    "gles2": "es3.0",
    "glsc2": "gl4",
}

_DOCBOOK_NS = "http://docbook.org/ns/docbook"


def _text_content(node: Node) -> str:
    """Recursively extract plain text from a DocBook node, stripping all markup."""
    parts: list[str] = []
    for child in node.childNodes:
        if child.nodeType == Node.TEXT_NODE:
            parts.append(child.nodeValue)
        elif child.nodeType == Node.ELEMENT_NODE:
            parts.append(_text_content(child))
    return "".join(parts)


def _strip_whitespace(text: str) -> str:
    """Collapse internal whitespace and strip leading/trailing."""
    return " ".join(text.split())


def _find_elements_by_local_name(parent: Node, local_name: str) -> list[Node]:
    """Find all direct child elements matching a local tag name (namespace-agnostic)."""
    results: list[Node] = []
    for child in parent.childNodes:
        if child.nodeType == Node.ELEMENT_NODE and child.localName == local_name:
            results.append(child)
    return results


def _find_element_by_local_name(parent: Node, local_name: str) -> Node | None:
    """Find the first direct child element matching a local tag name."""
    for child in parent.childNodes:
        if child.nodeType == Node.ELEMENT_NODE and child.localName == local_name:
            return child
    return None


def _find_deep(parent: Node, local_name: str) -> list[Node]:
    """Recursively find all descendant elements with the given local name."""
    results: list[Node] = []
    for child in parent.childNodes:
        if child.nodeType == Node.ELEMENT_NODE:
            if child.localName == local_name:
                results.append(child)
            results.extend(_find_deep(child, local_name))
    return results


def _parse_refpage(path: str) -> list[FunctionDoc] | None:
    """Parse a single refpage XML file, returning FunctionDoc(s) or None on error.

    Returns a list because a single refpage may document multiple related functions
    (e.g., glClear covers just glClear, but glGenBuffers covers glGenBuffers and
    glCreateBuffers).  All share the same brief and param docs.
    """
    try:
        doc = xml.dom.minidom.parse(path)
    except Exception:
        return None

    root = doc.documentElement

    # Extract brief from <refpurpose>
    purposes = _find_deep(root, "refpurpose")
    if not purposes:
        return None
    brief = _strip_whitespace(_text_content(purposes[0]))
    if not brief:
        return None

    # Extract parameter descriptions from <refsect1 xml:id="parameters">
    params: dict[str, str] = {}
    for sect in _find_deep(root, "refsect1"):
        sect_id = sect.getAttribute("xml:id") or sect.getAttribute("id") or ""
        if sect_id != "parameters":
            continue
        for varentry in _find_deep(sect, "varlistentry"):
            terms = _find_deep(varentry, "term")
            listitems = _find_deep(varentry, "listitem")
            if not terms or not listitems:
                continue
            # The parameter name is inside <term><parameter>name</parameter></term>
            param_nodes = _find_deep(terms[0], "parameter")
            if not param_nodes:
                continue
            param_name = _strip_whitespace(_text_content(param_nodes[0]))
            # The description is in the first <para> inside <listitem>
            paras = _find_deep(listitems[0], "para")
            if paras:
                desc = _strip_whitespace(_text_content(paras[0]))
                if desc:
                    params[param_name] = desc
        break  # Only process the first parameters section

    func_doc = FunctionDoc(brief=brief, params=params)

    # Collect all function names documented in this refpage
    names: list[str] = []
    for refname in _find_deep(root, "refname"):
        name = _strip_whitespace(_text_content(refname))
        if name:
            names.append(name)

    if not names:
        return None

    return [FunctionDoc(brief=func_doc.brief, params=dict(func_doc.params)) for _ in names], names


def parse_refpages(refpages_dir: str, api: str = "gl") -> dict[str, FunctionDoc]:
    """Parse all refpage XML files in the appropriate subdirectory.

    Args:
        refpages_dir: Path to the OpenGL-Refpages root directory.
        api: The Glaze API name (e.g., 'gl', 'gles2').

    Returns:
        A dict mapping GL function names (e.g., 'glClear') to their FunctionDoc.
    """
    subdir = _API_TO_REFDIR.get(api, "gl4")
    search_dir = os.path.join(refpages_dir, subdir)
    if not os.path.isdir(search_dir):
        return {}

    index: dict[str, FunctionDoc] = {}
    for filename in os.listdir(search_dir):
        if not filename.endswith(".xml") or not filename.startswith("gl"):
            continue
        filepath = os.path.join(search_dir, filename)
        result = _parse_refpage(filepath)
        if result is None:
            continue
        docs, names = result
        for doc, name in zip(docs, names, strict=True):
            index[name] = doc

    return index

File: conan/packages/glaze-cpp/src/libglaze-docparser/src/glaze/docparser/DocParser.cpp
#include "glaze/docparser/DocParser.h"

#include <pugixml.hpp>

#include <string>
#include <string_view>
#include <system_error>
#include <unordered_map>
#include <vector>

namespace glaze::docparser {

namespace {

//! Mirror of `_API_TO_REFDIR` in `doc_parser.py`. Unknown APIs fall back to
//! "gl4" which is the most complete refpage set.
std::string refpagesSubdir(std::string_view api) {
    if (api == "gles1") {
        return "es1.1";
    }
    if (api == "gles2") {
        return "es3.0";
    }
    // gl, gl_compat, glsc2, or anything else -> gl4
    return "gl4";
}

//! Extract every text descendant concatenated — matches Python
//! `_text_content` in `doc_parser.py`.
void collectText(const pugi::xml_node &node, std::string &out) {
    for (auto child = node.first_child(); child; child = child.next_sibling()) {
        switch (child.type()) {
        case pugi::node_pcdata:
        case pugi::node_cdata:
            out.append(child.value());
            break;
        case pugi::node_element:
            collectText(child, out);
            break;
        default:
            break;
        }
    }
}

std::string textOf(const pugi::xml_node &node) {
    std::string buf;
    collectText(node, buf);
    return buf;
}

//! Collapse internal whitespace, strip leading/trailing. Mirrors Python
//! `_strip_whitespace` which does `" ".join(text.split())`.
std::string stripWhitespace(std::string_view text) {
    std::string out;
    out.reserve(text.size());
    bool first = true;
    for (std::size_t i = 0; i < text.size();) {
        // skip whitespace runs
        while (i < text.size() && (text[i] == ' ' || text[i] == '\t' ||
                                   text[i] == '\n' || text[i] == '\r' ||
                                   text[i] == '\f' || text[i] == '\v')) {
            ++i;
        }
        std::size_t start = i;
        while (i < text.size() && !(text[i] == ' ' || text[i] == '\t' ||
                                    text[i] == '\n' || text[i] == '\r' ||
                                    text[i] == '\f' || text[i] == '\v')) {
            ++i;
        }
        if (start < i) {
            if (!first) {
                out.push_back(' ');
            }
            out.append(text, start, i - start);
            first = false;
        }
    }
    return out;
}

//! Recursively find descendant element nodes matching the given name.
//! pugixml doesn't handle XML namespaces specially — it preserves the raw
//! element names. The OpenGL-Refpages XMLs use a default namespace with no
//! prefix, so `child.name()` equals the local name directly.
void findDeepImpl(const pugi::xml_node &parent, const char *name,
                  std::vector<pugi::xml_node> &out) {
    for (auto child = parent.first_child(); child; child = child.next_sibling()) {
        if (child.type() != pugi::node_element) {
            continue;
        }
        if (std::string_view{child.name()} == name) {
            out.push_back(child);
        }
        findDeepImpl(child, name, out);
    }
}

std::vector<pugi::xml_node> findDeep(const pugi::xml_node &parent,
                                     const char *name) {
    std::vector<pugi::xml_node> results;
    findDeepImpl(parent, name, results);
    return results;
}

pugi::xml_node findFirst(const pugi::xml_node &parent, const char *name) {
    for (auto child = parent.first_child(); child; child = child.next_sibling()) {
        if (child.type() == pugi::node_element &&
            std::string_view{child.name()} == name) {
            return child;
        }
        auto deep = findFirst(child, name);
        if (deep) {
            return deep;
        }
    }
    return {};
}

//! Look up the `xml:id` or `id` attribute on a section. DocBook 5 uses
//! `xml:id` but the Python reference also accepts plain `id`.
std::string_view sectionId(const pugi::xml_node &sect) {
    if (auto attr = sect.attribute("xml:id")) {
        return attr.value();
    }
    if (auto attr = sect.attribute("id")) {
        return attr.value();
    }
    return {};
}

struct ParsedRefpage {
    std::vector<std::string> names;
    FunctionDoc doc;
};

//! Parse one refpage XML file. Returns std::nullopt-equivalent empty
//! ParsedRefpage (no names) when the file cannot be parsed or is missing
//! a brief. Matches the Python reference's "silently skip" behaviour.
ParsedRefpage parseRefpage(const std::filesystem::path &path) {
    ParsedRefpage result;
    pugi::xml_document doc;
    // Keep DOCTYPE intact: OpenGL-Refpages files use DOCTYPE ENTITY
    // references (e.g. mathent) which pugixml happily ignores by default.
    const auto load = doc.load_file(path.string().c_str(),
                                    pugi::parse_default | pugi::parse_ws_pcdata);
    if (!load) {
        return result;
    }
    auto root = doc.document_element();
    if (!root) {
        return result;
    }

    // 1. Brief — first <refpurpose> anywhere in the tree.
    const auto purpose = findFirst(root, "refpurpose");
    if (!purpose) {
        return result;
    }
    result.doc.brief = stripWhitespace(textOf(purpose));
    if (result.doc.brief.empty()) {
        return result;
    }

    // 2. Parameters — first <refsect1 xml:id="parameters">.
    for (const auto &sect : findDeep(root, "refsect1")) {
        if (sectionId(sect) != "parameters") {
            continue;
        }
        for (const auto &varentry : findDeep(sect, "varlistentry")) {
            const auto terms = findDeep(varentry, "term");
            const auto listitems = findDeep(varentry, "listitem");
            if (terms.empty() || listitems.empty()) {
                continue;
            }
            const auto paramNodes = findDeep(terms.front(), "parameter");
            if (paramNodes.empty()) {
                continue;
            }
            const auto paramName = stripWhitespace(textOf(paramNodes.front()));
            if (paramName.empty()) {
                continue;
            }
            const auto paras = findDeep(listitems.front(), "para");
            if (paras.empty()) {
                continue;
            }
            auto desc = stripWhitespace(textOf(paras.front()));
            if (!desc.empty()) {
                result.doc.params.emplace(paramName, std::move(desc));
            }
        }
        break; // only first <refsect1 id=parameters> like Python
    }

    // 3. Names — every <refname> declared in the refpage. All share the
    //    same brief/params payload.
    for (const auto &refname : findDeep(root, "refname")) {
        auto name = stripWhitespace(textOf(refname));
        if (!name.empty()) {
            result.names.push_back(std::move(name));
        }
    }
    return result;
}

} // namespace

DocIndex DocParser::parseRefpages(const std::filesystem::path &refpagesDir,
                                  std::string_view api) const {
    DocIndex index;
    std::error_code ec;
    const auto searchDir = refpagesDir / refpagesSubdir(api);
    if (!std::filesystem::is_directory(searchDir, ec)) {
        return index;
    }
    for (const auto &entry :
         std::filesystem::directory_iterator{searchDir, ec}) {
        if (ec) {
            break;
        }
        if (!entry.is_regular_file(ec)) {
            continue;
        }
        const auto &path = entry.path();
        if (path.extension() != ".xml") {
            continue;
        }
        const auto filename = path.filename().string();
        if (filename.rfind("gl", 0) != 0) {
            continue; // match Python `filename.startswith("gl")`
        }
        auto parsed = parseRefpage(path);
        if (parsed.names.empty()) {
            continue;
        }
        for (const auto &name : parsed.names) {
            // Each name gets its own copy — matches the Python behaviour
            // where all functions documented by the same refpage share
            // the same (deep-copied) brief and params.
            index.insert_or_assign(name, parsed.doc);
        }
    }
    return index;
}

} // namespace glaze::docparser
